helix-core/src/main/java/org/apache/helix/controller/GenericHelixController.java

```diff
@@ -828,12 +830,17 @@ public void onCustomizedStateRootChange(String instanceName, List<String> custom
     }
 
     // TODO: remove the synchronization here once we move this update into dataCache.
-    synchronized (_lastSeenCustomizedStateTypes) {
-      Set<String> lastSeenCustomizedStateTypes = _lastSeenCustomizedStateTypes.get();
+    synchronized (_lastSeenCustomizedStateTypesMap) {
+      Map<String, Set<String>> lastSeenCustomizedStateTypesMap =
+          _lastSeenCustomizedStateTypesMap.get();
+      Set<String> lastSeenCustomizedStateTypes = Collections.emptySet();
+      if (lastSeenCustomizedStateTypesMap != null && lastSeenCustomizedStateTypesMap
```

Comment:
Might help to simplify the code.

Map<String, Set<String>> lastSeenCustomizedStateTypesMap =
          _lastSeenCustomizedStateTypesMap.get();
if (null == lastSeenCustomizedStateTypesMap) {
  lastSeenCustomizedStateTypesMap = new HashMap();
  // lazy init the AtomicReference
  _lastSeenCustomizedStateTypesMapRef.set(lastSeenCustomizedStateTypesMap);
}
if (!lastSeenCustomizedStateTypesMap.containsKey(instanceName)) {
  lastSeenCustomizedStateTypesMap.put(instanceName, Collections.emptySet());
}
Set<String> lastSeenCustomizedStateTypes = lastSeenCustomizedStateTypesMap.get(instanceName);

Then you don't need to check null later.